docs/proposals/01-extensibility.md

```diff
@@ -0,0 +1,879 @@
+| botanist.WaitUntilKubeAddonManagerDeleted | Unchanged; Gardener waits until the kube-addon-manager is deleted. |
+| botanist.CleanCustomResourceDefinitions | Unchanged, Gardener cleans the CRDs in the Shoot. |
+| botanist.CleanKubernetesResources | Unchanged, Gardener cleans all remaining Kubernetes resources in the Shoot. |
+| hybridBotanist.DestroyMachines | Gardener deletes the WorkerPool-specific CRD in the Seed, and the responsible WorkerPool-controller picks it up and does its job.<br>Gardener waits until the CRD is deleted. |
+| botanist.DestroyIngressDNSRecord | Gardener deletes the DNS-specific CRD in the Seed, and the responsible DNS-controller picks it up and does its job. <br>Gardener waits until the CRD is deleted. |
+| shootCloudBotanist.DestroyKube2IAMResources | This function would disappear (currently Gardener would execute a Terraform job deleting the IAM roles specified in the `Shoot` manifest). We cannot keep this behavior, the user would be responsible to delete the needed IAM roles on its own. |
+| shootCloudBotanist.DestroyInfrastructure | Gardener deletes the Infrastructure-specific CRD in the Seed, and the responsible Botanist picks it up and does its job.<br>Gardener waits until the CRD is deleted. |
+| botanist.DestroyExternalDomainDNSRecord | Gardener deletes the DNS-specific CRD in the Seed, and the responsible DNS-controller picks it up and does its job.<br>Gardener waits until the CRD is deleted. |
+| botanist.DeleteKubeAPIServer | Unchanged; Gardener deletes the kube-apiserver. |
+| botanist.DeleteBackupInfrastructure | Unchanged; Gardener deletes the `BackupInfrastructure` object in the Garden cluster.<br>(The BackupInfrastructure controller deletes the BackupInfrastructure-specific CRD in the Seed, and the responsible Botanist picks it up and does its job.<br>The BackupInfrastructure controller waits until the CRD is deleted.) |
+| botanist.DestroyInternalDomainDNSRecord | Gardener deletes the DNS-specific CRD in the Seed, and the responsible DNS-controller picks it up and does its job.<br>Gardener waits until the CRD is deleted. |
+| botanist.DeleteNamespace | Unchanged; Gardener deletes the Shoot namespace in the Seed cluster. |
+| botanist.WaitUntilSeedNamespaceDeleted | Unchanged; Gardener waits until the Shoot namespace in the Seed has been deleted. |
+| botanist.DeleteGardenSecrets | Unchanged; Gardener deletes the kubeconfig/ssh-keypair `Secret` in the project namespace in the Garden. |
+
+### Gardenlet
+
+One part of the whole extensibility work will also to further split Gardener itself.
+Inspired from Kubernetes itself we plan to move the `Shoot` reconciliation/deletion controller loops as well as the `BackupInfrastructure` reconciliation/deletion controller loops into a dedicated "gardenlet" component that will run in the Seed cluster.
+With that, it can talk locally to the responsible kube-apiserver and we do no longer need to perform every operation out of the Garden cluster.
```

Comment:
What could be an option would be to let Gardener write a full/complete descriptive resource into the seed containing all the information needed for the gardenlet to create the shoot.

Reply:
As in the personal discussion, I am confused. If we build the core of the extensibility, control flow remains garden->seed->shoot with more/all that is today garden->shoot becoming seed->shoot. Ideally, GEP #1 brings already what you want @Adracus. Gardenlet is a refinement with another purpose: Gardenlet is @mvladev idea to support a Gardener out in the public and then let everybody speak to the Gardener, when that's not necessarily possible vice versa (e.g. from private cloud installations). For that purpose (IF we do that) the control flow has to change from garden->seed (pushing "tasks") to seed->garden (pulling "tasks").

Reply:
@rfranzke that's exactly what I was thinking of. @vlerenc I just see some big disadvantages with the flow reversal:
* Higher network load on the Gardener cluster: By reversing the flow, every seed will start polling from the Gardener cluster, effectively resulting in lots of requests to the API server, making this a potential bottleneck (I mean yes, you can scale horizontally, but network will be the problem here)
* No ability to reconcile without the Gardener cluster: By introducing this reversal, the Gardenlet will not work anymore with a broken network connection
* Gardenlet will have to communicate with *3 different K8s clusters*: The Gardener from which it polls, the one it runs in in order to deploy control plane components etc. + the resulting shoot cluster - this is just a whole lot of complexity in this component.
* No reuse potential: If we develop the Gardenlet with this architecture in mind, it will just make sense in the context of Gardener. However, if we develop it with the architecture that I proposed (full complete descriptive resource in the seed) this could also be used standalone (i.e. without the Garden cluster) to create one-off Shoots without Gardener management. Effectively, this comes very close / is an implementation of the cluster API.

Reply:
Again, I don't understand the discussion. IF we need/want to have only one Gardener per landscape, we will do that. It's not a local design decision. It's the fundamental question of how we will handle potentially many isolated Clouds (like SAP DCs) to avoid multiple unmanageable Gardener installations. This is the only thing that matters and will drive the decision. Of course, I am not/never blind to technical considerations, but there are none that would be relevant in the light of the effect the reversal of the control flow will have for us.

Still, let's talk about your points:
* API servers do the same; if we do the Gardenlets, we do nothing else than Kubelets and Kubernetes API servers do already; we should have no problems in scaling, too
* The point about Reconciliation, I completely fail to understand; if the Garden cluster is offline, Gardener will neither talk into the seed, nor the Gardenlet can talk to the Gardener; net effect is the same, nothing breaks, but nothing progresses either; Gardener is the source of truth, whether it pushes "tasks" or the Gardenlet pulls "tasks"
* 3 different K8s clusters is why an issue?
* Please elaborate on what you have proposed and why/where it differs from the extensibility plans (ignore the Gardenlet - that's for me a separate question, driven exclusively by the business need).